In [ ]:
import numpy as np

# Function to generate sequences of 1s and 2s
def generate_sequences(n_tosses, n_experiments):
    sequences = []
    for _ in range(n_experiments):
        sequence = np.random.choice([1, 2], size=n_tosses)
        sequences.append(sequence)
    return sequences

# Parameters
n_experiments = 1

# Generate sequences for 10, 50, and 100 tosses
tosses_list = [10, 50, 100]
sequences_dict = {}

for n_tosses in tosses_list:
    sequences = generate_sequences(n_tosses, n_experiments)
    sequences_dict[n_tosses] = sequences # Fixed indentation

# Display results
for n_tosses, sequences in sequences_dict.items(): # Added .items() and :
    print(f"Results for {n_tosses} tosses:")
    for sequence in sequences:
        print(sequence)
    print("\n")

Results for 10 tosses:
[1 2 2 2 2 2 2 1 1 1]


Results for 50 tosses:
[2 1 1 1 1 1 2 1 2 2 2 1 1 2 2 1 1 2 2 1 1 1 2 2 1 2 2 1 2 2 2 1 2 2 2 1 1
 2 2 2 2 2 2 1 1 2 1 1 1 1]


Results for 100 tosses:
[2 1 1 2 1 2 2 2 2 1 1 2 2 1 2 1 2 2 2 2 1 1 1 1 2 1 2 2 2 1 2 2 2 2 1 1 1
 2 1 2 2 1 2 1 1 2 2 1 1 1 1 1 1 1 1 2 1 2 1 2 1 1 2 1 1 1 1 2 1 2 1 1 1 1
 2 1 1 2 2 1 2 1 1 1 1 1 1 2 1 1 2 2 2 1 2 2 2 1 1 2]




**Example Result**

Results for 10 tosses:
[1 1 1 2 2 2 2 2 1 1]


Results for 50 tosses:
[2 2 2 2 2 2 2 1 1 1 2 1 2 1 1 1 1 1 2 2 1 1 2 1 1 2 1 2 2 2 1 2 1 2 2 2 2
 1 1 2 2 2 1 1 2 1 1 2 1 1]


Results for 100 tosses:
[2 1 2 1 2 2 1 2 2 2 1 2 1 1 2 1 2 1 1 2 1 1 2 1 1 2 2 2 1 1 2 1 1 1 2 1 1
 1 2 1 1 2 1 2 1 2 2 1 1 2 1 1 1 2 2 1 1 2 2 2 2 1 2 2 1 1 1 1 1 1 2 2 1 2
 2 2 1 1 1 2 1 1 2 1 2 1 2 2 2 2 2 2 2 2 1 1 1 1 2 2]

**Monte Carlo Simulation**
Computation

In [ ]:
import numpy as np

# Function to simulate coin toss
def coin_toss_simulation(n_tosses, n_experiments):
    results = []
    for _ in range(n_experiments):
        tosses = np.random.choice(['H', 'T'], size=n_tosses)
        heads_count = np.sum(tosses == 'H')
        results.append(heads_count)
    return results

# Function to calculate theoretical probability of exactly k heads
def theoretical_probability(n_tosses, k):
    from math import comb
    probability = comb(n_tosses, k) * (0.5 ** k) * (0.5 ** (n_tosses - k))
    return probability

# Function to calculate probability from simulation results
def calculate_probability(results, k):
    count_k_heads = results.count(k)
    probability = count_k_heads / len(results)
    return probability

# Parameters
n_experiments = 1000

# Simulate for 10, 50, and 100 tosses
tosses_list = [10, 50, 100]
results_dict = {}

for n_tosses in tosses_list:
    results = coin_toss_simulation(n_tosses, n_experiments)
    results_dict[n_tosses] = results

# Display results and calculate probabilities
for n_tosses, results in results_dict.items():
    print(f"Results for {n_tosses} tosses (first 10 outcomes):")
    print(results[:10])  # Display first 10 outcomes for brevity

    if n_tosses == 10:
        k = 5
    elif n_tosses == 50:
        k = 25
    elif n_tosses == 100:
        k = 50

    # Theoretical probability
    theoretical_prob = theoretical_probability(n_tosses, k)
    print(f"Theoretical probability of getting exactly {k} heads in {n_tosses} tosses: {theoretical_prob:.4f}")

    # Monte Carlo probability
    monte_carlo_prob = calculate_probability(results, k)
    print(f"Monte Carlo probability of getting exactly {k} heads in {n_tosses} tosses: {monte_carlo_prob:.4f}")
    print("\n")

Results for 10 tosses (first 10 outcomes):
[8, 3, 4, 7, 8, 4, 5, 5, 4, 8]
Theoretical probability of getting exactly 5 heads in 10 tosses: 0.2461
Monte Carlo probability of getting exactly 5 heads in 10 tosses: 0.2400


Results for 50 tosses (first 10 outcomes):
[22, 31, 28, 29, 25, 28, 22, 22, 31, 18]
Theoretical probability of getting exactly 25 heads in 50 tosses: 0.1123
Monte Carlo probability of getting exactly 25 heads in 50 tosses: 0.1140


Results for 100 tosses (first 10 outcomes):
[49, 48, 52, 54, 46, 54, 49, 48, 47, 57]
Theoretical probability of getting exactly 50 heads in 100 tosses: 0.0796
Monte Carlo probability of getting exactly 50 heads in 100 tosses: 0.0820




Example:

Results for 10 tosses (first 10 outcomes):
[5, 2, 6, 6, 4, 7, 6, 4, 2, 6]
Theoretical probability of getting exactly 5 heads in 10 tosses: 0.2461
Monte Carlo probability of getting exactly 5 heads in 10 tosses: 0.2290


Results for 50 tosses (first 10 outcomes):
[20, 21, 31, 22, 25, 21, 24, 29, 23, 22]
Theoretical probability of getting exactly 25 heads in 50 tosses: 0.1123
Monte Carlo probability of getting exactly 25 heads in 50 tosses: 0.1090


Results for 100 tosses (first 10 outcomes):
[52, 39, 50, 44, 54, 43, 47, 46, 54, 47]
Theoretical probability of getting exactly 50 heads in 100 tosses: 0.0796
Monte Carlo probability of getting exactly 50 heads in 100 tosses: 0.0840